crates/video/src/video_backends/v4l2_decoder.rs

```diff
@@ -0,0 +1,845 @@
+    }
+
+    fn query_control(&self, control: video::ControlType) -> video::CmdResponseType {
+        let (id, flags) = v4l2r::ioctl::parse_ctrl_id_and_flags(control.to_v4l2());
+        let queryctrl: v4l2_queryctrl = match v4l2r::ioctl::queryctrl(&self.video_device, id, flags)
+        {
+            Ok(queryctrl) => queryctrl,
+            Err(e) => {
+                warn!("queryctrl failed: {}", e);
+                return Sync(video::CmdResponse::Error(UnsupportedControl));
+            }
+        };
+
+        Sync(video::CmdResponse::QueryControl(
+            video::VideoControl::new_from_type(control, queryctrl.type_.into()),
+        ))
+    }
+
+    fn clear_queue(
+        &mut self,
+        stream_id: u32,
+        queue_type: video::QueueType,
+    ) -> video::CmdResponseType {
+        let stream = match self.streams.get_mut(&stream_id) {
+            Some(stream) => stream,
+            None => {
+                return Sync(video::CmdResponse::Error(InvalidStreamId));
+            }
+        };
+        let queue = v4l2r::QueueType::from_value(queue_type as u32).unwrap();
+        // send replies for all in-flight buffers
+        for resource in stream.queued_resources_mut(queue_type) {
+            resource.ready_with(video::BufferFlags::ERR, 0);
+        }
+        /*
+         * QUEUE_CLEAR behaviour from virtio-video spec
+         * Return already queued buffers back from the input or the output queue
+         * of the device. The device SHOULD return all of the buffers from the
+         * respective queue as soon as possible without pushing the buffers through
+         * the processing pipeline.
+         *
+         * From v4l2 PoV we issue a VIDIOC_STREAMOFF on the queue which will abort
+         * or finish any DMA in progress, unlocks any user pointer buffers locked
+         * in physical memory, and it removes all buffers from the incoming and
+         * outgoing queues.
+         */
+        if let Err(e) = v4l2r::ioctl::streamoff(stream, queue) {
+            warn!("streamoff failed: {}", e);
+            return Sync(video::CmdResponse::Error(InvalidParameter));
+        }
+
+        Sync(video::CmdResponse::OkNoData)
+    }
+
+    fn get_params(&self, stream_id: u32, queue_type: video::QueueType) -> video::CmdResponseType {
+        let stream = match self.streams.get(&stream_id) {
+            Some(stream) => stream,
+            None => {
+                return Sync(video::CmdResponse::Error(InvalidStreamId));
+            }
+        };
+        let queue = v4l2r::QueueType::from_value(queue_type as u32).unwrap();
+        let mut params = video::virtio_video_params::default();
+        let format: v4l2_format = match v4l2r::ioctl::g_fmt(stream, queue) {
+            Ok(format) => format,
+            Err(e) => {
+                warn!("g_fmt failed: {}", e);
+                return Sync(video::CmdResponse::Error(InvalidParameter));
+            }
+        };
```

Comment:
no particular preference on this, but this could probably be a `.map_err(...)?`?

Reply:
It leaves (in this case) an unused `Result`, so it would need a `let _ =` at the beginning. Same with most cases in this file.

Either way works for me, I've no particular preference, but I feel the current `if let Err(e) = ` looks a bit more explicit.

Reply:
Hm... Not sure why it would be unused? If you use `?` on it, it should not complain about that? Or am I misunderstanding the problem?

Reply:
It returns a `video::CmdResponseType`, not an Error. The response is what ultimately will get written in the vring back to the driver. In this case the response indicates that an error has occurred, but is a valid response, so I do not want to use `?`.